sbat.c

```diff
@@ -294,4 +294,96 @@ parse_sbat_var(list_t *entries)
 	return parse_sbat_var_data(entries, data, datasize+1);
 }
 
+EFI_STATUS
+set_sbat_uefi_variable(void)
+{
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	UINT32 attributes = 0;
+
+	CHAR8 *sbat = NULL;
+	UINTN sbatsize = 0;
+
+	/* get size */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	if (EFI_ERROR(efi_status))
+		dprint(L"SBAT read failed %r\n", efi_status);
+
+	sbat = AllocateZeroPool(sbatsize);
+	if (!sbat)
+		return EFI_OUT_OF_RESOURCES;
+
+	/* read and get attributes */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	if (EFI_ERROR(efi_status))
+		dprint(L"SBAT read failed %r\n", efi_status);
+
+	/*
+	 * If attributes and the initial part of the version match, then
+	 * there is no need to re-write it. If it is not set or the read
+	 * failed, then attributes won't match.
+	 *
+	 * Finally, if the SBAT_VAR built into this shim is newer, update
+	 * it. Either the sbat version or date can be newer, and since
+	 * the version is more significant than the date, strcmp() applies.
+	 */
+	if (((attributes != (EFI_VARIABLE_NON_VOLATILE |
+		EFI_VARIABLE_BOOTSERVICE_ACCESS))
+		|| (attributes != (EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)))
+		|| (strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) != 0)
+		|| (strncmp(SBAT_VAR, sbat, strlen(SBAT_VAR)) > 0)
+		) {
```

Comment:
You probably should consider adding something like the following before this function:
```
#define NV_BS (EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS)
#define NV_BS_TIMEAUTH \
        (NV_BS | EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)
```
And then making this be:
```
        /*
         * lorem ipsum ...
         */
        if (!EFI_ERROR(efi_status) &&
            (attributes != NV_BS || attributes != NV_BS_TIMEAUTH ||
             strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) == 0 ||
             strncmp(sbat, SBAT_VAR, strlen(SBAT_VAR)) <= 0)) {
                FreePool(sbat);
                return EFI_SUCCESS;
        }
        ...
```

Reply:
Doing the cleanup, but I do think we want to drive on into this if there's an error status on the read. Best case we succeed and set the variable despite it. Worst case we fail, which is the same as giving up.

Reply:
Would something like this be clearer to you:
```
        if (EFI_ERROR(efi_status)) {
                dprint(L"SBAT read failed %r\n", efi_status);
        } else if ((attributes == NV_BS || attributes == NV_BS_TIME_AUTH) &&
                   (strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) == 0 ||
                    strncmp(sbat, SBAT_VAR, strlen(SBAT_VAR)) <= 0)) {
                FreePool(sbat);
                return EFI_SUCCESS;
        }

```

Reply:
(I think I've got the negation right that time, but certainly double check it.)

Reply:
Yes, that one is right. Thank you.